songpal/device.py

```diff
@@ -130,31 +139,68 @@ async def get_supported_methods(self):
         Calling this as the first thing before doing anything else is
         necessary to fill the available services table.
         """
-        response = await self.request_supported_methods()
+        try:
+            response = await self.request_supported_methods()
 
-        if "result" in response:
-            services = response["result"][0]
-            _LOGGER.debug("Got %s services!" % len(services))
+            if "result" in response:
+                services = response["result"][0]
+                _LOGGER.debug("Got %s services!" % len(services))
 
-            for x in services:
-                serv = await Service.from_payload(
-                    x, self.endpoint, self.idgen, self.debug, self.force_protocol
-                )
-                if serv is not None:
-                    self.services[x["service"]] = serv
-                else:
-                    _LOGGER.warning("Unable to create service %s", x["service"])
+                for x in services:
+                    serv = await Service.from_payload(
+                        x, self.endpoint, self.idgen, self.debug, self.force_protocol
+                    )
+                    if serv is not None:
+                        self.services[x["service"]] = serv
+                    else:
+                        _LOGGER.warning("Unable to create service %s", x["service"])
 
-            for service in self.services.values():
-                if self.debug > 1:
-                    _LOGGER.debug("Service %s", service)
-                for api in service.methods:
-                    # self.logger.debug("%s > %s" % (service, api))
+                for service in self.services.values():
                     if self.debug > 1:
-                        _LOGGER.debug("> %s" % api)
+                        _LOGGER.debug("Service %s", service)
+                    for api in service.methods:
+                        # self.logger.debug("%s > %s" % (service, api))
+                        if self.debug > 1:
+                            _LOGGER.debug("> %s" % api)
+                return self.services
+
+            return None
+        except SongpalException as e:
+            found_services = None
+            if e.code == 12 and e.error_message == "getSupportedApiInfo":
+                found_services = await self._get_supported_methods_upnp()
```

Comment:
Please separate all UPnP functionality into a separate class.

Reply:
Some UPnP  code is also inside the containers (such as input & volume), how would you recommend separating this?
Having separate UPnP containers that subclass the regular container?

Reply:
Yeah, I think separate classes implementing the same interface is the way to go. Maybe something like `VolumeControl`, `InputControl`, which provide the common interface that gets implemented either by songpal or upnp impl, which gets chosen during the initialization procedures?

The code is quite tangled at the moment, especially on the containers and the device files, which was one of the reasons I started refactoring towards service-based classes.